### DISCUSSION TITLE
How to influence whitespaces with @babel/generator

### DISCUSSION THREAD

**Discussion by bali182** [babel/babel#13479] [Q&A]
I'm generating source code from ASTs (typescript in my case) using @babel/generator. Consider the following code:

```ts
import {
  tsTypeAliasDeclaration,
  identifier,
  tsTypeLiteral,
  tsPropertySignature,
  tsTypeAnnotation,
  tsStringKeyword,
  program,
} from '@babel/types'
import generate from '@babel/generator'

describe('Generating with @babel/types', () => {
  it('Should make 2 types and print them', () => {
    const firstTypeAst = tsTypeAliasDeclaration(
      identifier('Foo'),
      undefined,
      tsTypeLiteral([tsPropertySignature(identifier('foo'), tsTypeAnnotation(tsStringKeyword()))]),
    )
    const secondTypeAst = tsTypeAliasDeclaration(
      identifier('Bar'),
      undefined,
      tsTypeLiteral([tsPropertySignature(identifier('bar'), tsTypeAnnotation(tsStringKeyword()))]),
    )
    const ast = program([firstTypeAst, secondTypeAst], [], 'module')
    const { code } = generate(ast)
    console.log(code)
  })
})
```

This results in the following generated code:

```ts
type Foo = {
  foo: string;
};
type Bar = {
  bar: string;
};
```

**How do I influence whitespaces?** Let's say I'd like to have a newline between the 2 type declarations:

```ts
type Foo = {
  foo: string;
};

type Bar = {
  bar: string;
};
```

Or remove the newlines in type literal:

```ts
type Foo = { foo: string; };
type Bar = { bar: string; };
```

Is this something I can do on the generator level or this has to be part of the AST? I'm interested in any solution!

**Comment by hzoo** [comment]
I would recommend using https://github.com/codemod-js/codemod/blob/master/packages/cli/README.md. Since we removed most formatting options in generator as it's not the main usecase, recast/prettier is better used for this. An ast wouldn't contain the whitespace information
